src/Compilers/Shared/BuildClient.cs

```diff
@@ -167,6 +167,24 @@ public Task<RunCompilationResult> RunCompilationAsync(IEnumerable<string> origin
         {
             BuildResponse buildResponse;
 
+            IDisposable npcs = null;
+            try 
+            {
+                // Mono configurations without named pipe support will throw a PNSE at some point in this process.
+                npcs = new System.IO.Pipes.NamedPipeClientStream(".", "nonexistentpipe", System.IO.Pipes.PipeDirection.InOut);
+                npcs.Dispose();
+            } 
+            catch (PlatformNotSupportedException) 
+            {
+                if (npcs != null) 
+                {
+                    // Compensate for broken finalizer in older builds of mono
+                    GC.SuppressFinalize(npcs);
+                }
+                Console.WriteLine("WARNING: /shared was requested but this build of mono does not support named pipes.");
```

Comment:
We should write to the log vs. the console here. The compiler should only write to the console when issuing diagnostics that are specifically formatted as a compiler diagnostic. This is because MSBuild will parse all output from the compiler and if it's not in a specific form the message will be possibly mis-interperpreted. Today when we can't use the server we fail silently and issue diagnostics in the log think we should do that here as it's the same scenario.